File: data/src/main/java/com/marvelcomics/brito/data/datasource/remote/utils/AuthHashGenerator.java
package com.marvelcomics.brito.data.datasource.remote.utils;

import com.marvelcomics.brito.infrastructure.exception.MarvelApiException;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class AuthHashGenerator {
    public String generateHash(String timestamp, String publicKey, String privateKey)
            throws MarvelApiException {
        try {
            String value = timestamp + privateKey + publicKey;
            MessageDigest md5Encoder = MessageDigest.getInstance("MD5");
            byte[] md5Bytes = md5Encoder.digest(value.getBytes());

            StringBuilder md5 = new StringBuilder();
            for (byte md5Byte : md5Bytes) {
                md5.append(Integer.toHexString((md5Byte & 0xFF) | 0x100).substring(1, 3));
            }
            return md5.toString();
        }catch (NoSuchAlgorithmException e) {
            throw new MarvelApiException("cannot generate the api key", e);
        }
    }
}

File: data/src/main/java/com/marvelcomics/brito/data/datasource/remote/utils/AuthHashGenerator.kt
package com.marvelcomics.brito.data.datasource.remote.utils

import com.marvelcomics.brito.infrastructure.exception.MarvelApiException
import java.math.BigInteger
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

class AuthHashGenerator {

    @Throws(MarvelApiException::class)
    fun generateHash(timestamp: String, publicKey: String, privateKey: String): String {
        return try {
            val value = timestamp + privateKey + publicKey
            val md5Encoder = MessageDigest.getInstance("MD5")
            val md5Bytes = md5Encoder.digest(value.toByteArray())
            BigInteger(1, md5Bytes).toString(16).padStart(32, '0')
        } catch (e: NoSuchAlgorithmException) {
            throw MarvelApiException("cannot generate the api key", e)
        }
    }
}